[JIT] Bind AliasInfo to decrease differences in interfaces across languages

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79814
* __->__ #79661
* #79474

- Changes python bindings to derive from AliasInfo class instead of Argument Class so that the C++ and python interfaces are more uniform

Commit:
[JIT] Bind AliasInfo to decrease differences in interfaces across languages

[ghstack-poisoned]

Commit:
Update on "[JIT] Bind AliasInfo to decrease differences in interfaces across languages"


- Changes python bindings to derive from AliasInfo class instead of Argument Class so that the C++ and python interfaces are more uniform

[ghstack-poisoned]

albanD commented on test/test_public_bindings.py:
```diff
@@ -33,6 +33,7 @@ def test_no_new_bindings(self):
         torch_C_allowlist_superset = {
             "AggregationType",
             "AliasDb",
+            "AliasInfo"
```
Do we really want `torch.AliasInfo` to be a thing? Is this new public API properly documented?

davidberard98 replied:
changed it to torch._C._AliasInfo so this won't be an issue

davidberard98 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1490,42 +1490,32 @@ void initJITBindings(PyObject* module) {
+      .def_property_readonly("kwarg_only", [](Argument& self) -> bool {
+        return self.kwarg_only();
+      });
+  py::class_<AliasInfo>(m, "AliasInfo")
```
You can probably make this "_AliasInfo" and then you won't need to add it into the torch_C_allowlist_superset

Commit:
Update on "[JIT] Bind AliasInfo to decrease differences in interfaces across languages"


- Changes python bindings to derive from AliasInfo class instead of Argument Class so that the C++ and python interfaces are more uniform

[ghstack-poisoned]

Commit:
Update on "[JIT] Bind AliasInfo to decrease differences in interfaces across languages"


- Changes python bindings to derive from AliasInfo class instead of Argument Class so that the C++ and python interfaces are more uniform

[ghstack-poisoned]

Commit:
Update on "[JIT] Bind AliasInfo to decrease differences in interfaces across languages"


- Changes python bindings to derive from AliasInfo class instead of Argument Class so that the C++ and python interfaces are more uniform

[ghstack-poisoned]

davidberard98 reviewed (approved):
LGTM!
